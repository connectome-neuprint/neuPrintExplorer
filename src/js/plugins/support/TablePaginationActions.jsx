import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import IconButton from '@mui/material/IconButton';
import FirstPageIcon from '@material-ui/icons/FirstPage';
import KeyboardArrowLeft from '@material-ui/icons/KeyboardArrowLeft';
import KeyboardArrowRight from '@material-ui/icons/KeyboardArrowRight';
import LastPageIcon from '@material-ui/icons/LastPage';

const styles = theme => ({
  root: {
    flexShrink: 0,
    color: theme.palette.text.secondary,
    marginLeft: theme.spacing(2.5),
  },
});

class TablePaginationActions extends React.Component {
  handleFirstPageButtonClick = event => {
    const { onPageChange } = this.props;
    onPageChange(event, 0);
  };

  handleBackButtonClick = event => {
    const { onPageChange, page } = this.props;
    onPageChange(event, page - 1);
  };

  handleNextButtonClick = event => {
    const { onPageChange, page } = this.props;
    onPageChange(event, page + 1);
  };

  handleLastPageButtonClick = event => {
    const { onPageChange, count, rowsPerPage } = this.props;
    onPageChange(
      event,
      Math.max(0, Math.ceil(count / rowsPerPage) - 1)
    );
  };

  render() {
    const { classes, count, page, rowsPerPage, theme } = this.props;

    return (
      <div className={classes.root}>
        <IconButton
          onClick={this.handleFirstPageButtonClick}
          disabled={page === 0}
          aria-label="First Page"
        >
          {theme.direction === 'rtl' ? <LastPageIcon /> : <FirstPageIcon />}
        </IconButton>
        <IconButton
          onClick={this.handleBackButtonClick}
          disabled={page === 0}
          aria-label="Previous Page"
        >
          {theme.direction === 'rtl' ? <KeyboardArrowRight /> : <KeyboardArrowLeft />}
        </IconButton>
        <IconButton
          onClick={this.handleNextButtonClick}
          disabled={page >= Math.ceil(count / rowsPerPage) - 1}
          aria-label="Next Page"
        >
          {theme.direction === 'rtl' ? <KeyboardArrowLeft /> : <KeyboardArrowRight />}
        </IconButton>
        <IconButton
          onClick={this.handleLastPageButtonClick}
          disabled={page >= Math.ceil(count / rowsPerPage) - 1}
          aria-label="Last Page"
        >
          {theme.direction === 'rtl' ? <FirstPageIcon /> : <LastPageIcon />}
        </IconButton>
      </div>
    );
  }
}

TablePaginationActions.propTypes = {
  classes: PropTypes.object.isRequired,
  count: PropTypes.number.isRequired,
  onPageChange: PropTypes.func.isRequired,
  page: PropTypes.number.isRequired,
  rowsPerPage: PropTypes.number.isRequired,
  theme: PropTypes.object.isRequired
};

export default withStyles(styles, { withTheme: true })(
  TablePaginationActions
);
