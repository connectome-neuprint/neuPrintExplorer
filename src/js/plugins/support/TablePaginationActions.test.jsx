import React from 'react';
import { render, screen } from '../../tests/test-utils';
// eslint-disable-next-line import/no-unresolved
import '@testing-library/jest-dom';
import 'regenerator-runtime/runtime';
import TablePaginationActions from './TablePaginationActions';

function doNothing() {}

describe('TablePaginationActions', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <TablePaginationActions count={50} onPageChange={doNothing} rowsPerPage={5} page={0} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('TablePaginationActions shows next link disabled', async () => {
    await render(
      <TablePaginationActions count={50} onPageChange={doNothing} rowsPerPage={5} page={10} />
    );

    const buttons = await screen.getAllByRole('button');
    expect(buttons[0]).not.toBeDisabled(); // First page button
    expect(buttons[3]).toBeDisabled(); // Last page button
  });

  test('TablePaginationActions shows prev link disabled', async () => {

    await render(
      <TablePaginationActions count={50} onPageChange={doNothing} rowsPerPage={5} page={0} />
    );

    const buttons = await screen.getAllByRole('button');
    expect(buttons[0]).toBeDisabled(); // First page button
    expect(buttons[3]).not.toBeDisabled(); // Last page button
  });

});
/*
test('TablePaginationActions shows alternate icons for rtl theme', () => {
  const theme = {
    direction: 'rtl',
  };
  const component = renderer.create(
    <TablePaginationActions
      count={50}
      onPageChange={doNothing}
      rowsPerPage={5}
      page={1}
      theme={theme}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('clicking on last page link sends user to last page', () => {
  const onPageChange = sinon.spy();
  const raw = (
    <TablePaginationActions count={50} onPageChange={onPageChange} rowsPerPage={5} page={0} />
  );
  const wrapper = mount(raw);
  wrapper.find('button').at(3).simulate('click');
  expect(onPageChange.calledOnce).toBe(true);
  expect(onPageChange.args[0][1]).toEqual(9);
});

test('clicking on first page link sends user to last page', () => {
  const onPageChange = sinon.spy();
  const raw = (
    <TablePaginationActions count={50} onPageChange={onPageChange} rowsPerPage={5} page={5} />
  );
  const wrapper = mount(raw);
  wrapper.find('button').at(0).simulate('click');
  expect(onPageChange.calledOnce).toBe(true);
  expect(onPageChange.args[0][1]).toEqual(0);
});

test('clicking on next page link sends user to next page', () => {
  const page = 4;
  const onPageChange = sinon.spy();
  const raw = (
    <TablePaginationActions count={50} onPageChange={onPageChange} rowsPerPage={5} page={page} />
  );
  const wrapper = mount(raw);
  wrapper.find('button').at(2).simulate('click');
  expect(onPageChange.calledOnce).toBe(true);
  expect(onPageChange.args[0][1]).toEqual(page + 1);
});

test('clicking on prev page link sends user to prev page', () => {
  const page = 4;
  const onPageChange = sinon.spy();
  const raw = (
    <TablePaginationActions count={50} onPageChange={onPageChange} rowsPerPage={5} page={page} />
  );
  const wrapper = mount(raw);
  wrapper.find('button').at(1).simulate('click');
  expect(onPageChange.calledOnce).toBe(true);
  expect(onPageChange.args[0][1]).toEqual(page - 1);
});
*/
