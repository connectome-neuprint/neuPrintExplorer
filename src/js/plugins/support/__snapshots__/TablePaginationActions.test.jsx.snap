// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePaginationActions renders correctly 1`] = `
<DocumentFragment>
  <div
    class="TablePaginationActions-root-1"
  >
    <button
      aria-label="First Page"
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-154kngp-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="FirstPageIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
        />
      </svg>
    </button>
    <button
      aria-label="Previous Page"
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-154kngp-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="KeyboardArrowLeftIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6z"
        />
      </svg>
    </button>
    <button
      aria-label="Next Page"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-154kngp-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="KeyboardArrowRightIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6z"
        />
      </svg>
    </button>
    <button
      aria-label="Last Page"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-154kngp-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="LastPageIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
